### PR TITLE
Added message-tags support

### PR DESCRIPTION
**Pull request by MrIron-no**

### CODE REVIEW

**Comment by hiddn** [comment]
I reviewed this PR and found two correctness issues, plus put together a small test for the new tag-parsing helpers. Posting as a suggested patch (I don't have push access to this fork) — feel free to take it as-is or adapt it.

**1. `inputCharBuffer` overflow risk** (`src/server_connection.cc`, `OnRead`)

`OnRead` copies incoming lines into the fixed-size `inputCharBuffer` via `strncpy(inputCharBuffer, line.c_str(), len + 1)` with no check that the line actually fits. `Buffer::ReadLine()` has no length cap of its own, so this was already a latent overflow, but this PR's own `IRC_MAX_MESSAGE` sizing (8703 bytes, up from 1024) makes long lines an expected case rather than a rare corruption on the wire — worth closing now rather than later. The fix below rejects and logs oversized lines instead of copying them.

**2. Silent default flip of TLS fingerprint burst** (`include/gnuworld_config.h`, `libircu/msg_N.cc`)

`NO_FINGERPRINT_BURST` (shipped `#undef`'d → fingerprint burst compiled in by default) got folded into `NEW_IRCU_FEATURES` (shipped `#undef`'d → compiled out by default). That's a reasonable consolidation since both gate "requires a modern ircu uplink", but it silently disables fingerprint burst for anyone who rebuilds without defining `NEW_IRCU_FEATURES`. I didn't touch the logic — just made the change loud instead of silent via a comment, so whoever merges/documents this can decide if that default is intended.

**3. Test coverage**

Added `test/xparameters_tags.cc` (wired into `test/Makefile.am` the same way as the other `noinst_PROGRAMS`) covering `escapeTagValue`/`unescapeTagValue` round-trips and edge cases (trailing `\`, invalid escapes), `parseTags` (basic, multi-tag, malformed/stray `;`), `formatTagPrefix`, and `hasTag`/`getTag`. All pass.

I verified: the two touched translation units compile cleanly in both the default and `-DNEW_IRCU_FEATURES` configs, and the new test binary builds and passes (compiled directly against `xparameters.h`, since this checkout wasn't configured with `--with-test`).

```diff
diff --git a/include/gnuworld_config.h b/include/gnuworld_config.h
index 62993a1b..18c7b92b 100644
--- a/include/gnuworld_config.h
+++ b/include/gnuworld_config.h
@@ -103,6 +103,14 @@
  * - TLS fingerprint burst (N ... +z)
  * - Outbound IRCv3 message-tags on P10
  * - CService: AC updates, SASL, network config, autohide (+x), etc.
+ *
+ * UPGRADE NOTE: TLS fingerprint burst used to be gated by its own flag,
+ * NO_FINGERPRINT_BURST, which shipped undefined and therefore compiled
+ * fingerprint burst in by default. It is now folded into this flag
+ * instead, which ships undefined (disabled) by default. Deployments
+ * that rely on TLS fingerprint burst and are rebuilding from a config
+ * predating this change must explicitly define NEW_IRCU_FEATURES to
+ * keep that behavior.
  */
 #undef NEW_IRCU_FEATURES
 
diff --git a/src/server_connection.cc b/src/server_connection.cc
index b5fee92c..225d74db 100644
--- a/src/server_connection.cc
+++ b/src/server_connection.cc
@@ -173,6 +173,12 @@ void xServer::OnRead(Connection* theConn, const string& line) {
         --len;
     }
 
+    if (len + 1 >= sizeof(inputCharBuffer)) {
+        elog << "xServer::OnRead> Rejecting oversized line (" << (len + 1) << " bytes, max "
+             << (sizeof(inputCharBuffer) - 1) << ")" << endl;
+        return;
+    }
+
     memset(inputCharBuffer, 0, sizeof(inputCharBuffer));
     strncpy(inputCharBuffer, line.c_str(), len + 1);
 
diff --git a/test/Makefile.am b/test/Makefile.am
index b1ca83c2..ce684399 100644
--- a/test/Makefile.am
+++ b/test/Makefile.am
@@ -12,7 +12,8 @@ noinst_PROGRAMS = test_burst \
 	test_signal \
 	test_match \
 	test_bot \
-	test_kick_transaction
+	test_kick_transaction \
+	test_xparameters_tags
 
 if COND_PCRE
 noinst_PROGRAMS += test_regex
@@ -61,6 +62,11 @@ test_kick_transaction_CXXFLAGS = -I${top_srcdir}/include \
 	-I${top_srcdir}/libgnuworld
 test_kick_transaction_LDADD = libgnuworld.la
 
+test_xparameters_tags_SOURCES = test/xparameters_tags.cc
+test_xparameters_tags_CXXFLAGS = -I${top_srcdir}/include \
+	-I${top_srcdir}/libgnuworld
+test_xparameters_tags_LDADD = libgnuworld.la
+
 test_regex_SOURCES = test/test_regex.cc
 test_regex_LDADD = libgnuworld.la -lpcre
 
diff --git a/test/xparameters_tags.cc b/test/xparameters_tags.cc
new file mode 100644
index 00000000..a311c792
--- /dev/null
+++ b/test/xparameters_tags.cc
@@ -0,0 +1,166 @@
+/**
+ * xparameters_tags.cc
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
+ * USA.
+ *
+ * Exercises the IRCv3 message-tag helpers added to xParameters:
+ * escapeTagValue/unescapeTagValue, parseTags, formatTagPrefix,
+ * hasTag/getTag.
+ */
+
+#include <iostream>
+#include <string>
+
+#include "xparameters.h"
+
+using namespace std;
+using namespace gnuworld;
+
+static int failures = 0;
+
+static void check(bool condition, const string& description) {
+    if (!condition) {
+        cout << "FAILED: " << description << endl;
+        ++failures;
+    }
+}
+
+static void testEscapeRoundTrip() {
+    // Every special character should survive escape -> unescape.
+    const string raw = "a;b c\\d\re\nf";
+    const string escaped = xParameters::escapeTagValue(raw);
+    check(escaped == "a\\:b\\sc\\\\d\\re\\nf",
+          "escapeTagValue produces the expected wire form");
+    check(xParameters::unescapeTagValue(escaped) == raw,
+          "unescapeTagValue reverses escapeTagValue");
+}
+
+static void testUnescapeEdgeCases() {
+    // Lone trailing backslash: produces no output character for it.
+    check(xParameters::unescapeTagValue("abc\\") == "abc",
+          "trailing lone backslash is dropped");
+
+    // Invalid escape sequence: drop the backslash, keep the character.
+    check(xParameters::unescapeTagValue("a\\bc") == "abc",
+          "invalid escape drops backslash and keeps following character");
+
+    // Empty value.
+    check(xParameters::unescapeTagValue("") == "", "empty value unescapes to empty");
+}
+
+static void testParseTagsBasic() {
+    xParameters::tagListType tags;
+
+    xParameters::parseTags("key=value", tags);
+    check(tags.size() == 1, "single key=value tag parses to one entry");
+    check(tags.size() == 1 && tags[0].key == "key" && tags[0].value == "value",
+          "single key=value tag has expected key/value");
+
+    xParameters::parseTags("flag", tags);
+    check(tags.size() == 1 && tags[0].key == "flag" && tags[0].value.empty(),
+          "key-only tag (no '=') parses with empty value");
+
+    xParameters::parseTags("a=1;b=2;c", tags);
+    check(tags.size() == 3, "multiple tags separated by ';' all parse");
+    check(tags.size() == 3 && tags[0].key == "a" && tags[0].value == "1" && tags[1].key == "b" &&
+              tags[1].value == "2" && tags[2].key == "c" && tags[2].value.empty(),
+          "multiple tags parse in order with correct key/value pairs");
+}
+
+static void testParseTagsMalformed() {
+    xParameters::tagListType tags;
+
+    // Stray/trailing ';' should not produce empty-key entries.
+    xParameters::parseTags(";a=1;;b=2;", tags);
+    check(tags.size() == 2 && tags[0].key == "a" && tags[1].key == "b",
+          "empty segments from stray/leading/trailing ';' are skipped");
+
+    xParameters::parseTags("", tags);
+    check(tags.empty(), "empty tag string parses to no tags");
+}
+
+static void testEscapedValueRoundTripThroughFormat() {
+    xParameters::tagListType tags;
+    tags.push_back(MessageTag{"msg", "hello; world \\ done"});
+    tags.push_back(MessageTag{"empty", ""});
+
+    const string prefix = xParameters::formatTagPrefix(tags);
+    check(!prefix.empty() && prefix.front() == '@' && prefix.back() == ' ',
+          "formatTagPrefix wraps tags with leading '@' and trailing space");
+
+    // Strip the leading '@' and trailing ' ' before re-parsing, mirroring
+    // what xServer::Process does on incoming lines.
+    const string tagBody = prefix.substr(1, prefix.size() - 2);
+
+    xParameters::tagListType roundTripped;
+    xParameters::parseTags(tagBody, roundTripped);
+
+    check(roundTripped.size() == 2, "round-tripped tag list has the same number of tags");
+    check(roundTripped.size() == 2 && roundTripped[0].key == "msg" &&
+              roundTripped[0].value == "hello; world \\ done",
+          "escaped value round-trips through formatTagPrefix/parseTags unchanged");
+    check(roundTripped.size() == 2 && roundTripped[1].key == "empty" &&
+              roundTripped[1].value.empty(),
+          "empty-value tag round-trips as a key with no '='");
+}
+
+static void testFormatTagPrefixEmpty() {
+    xParameters::tagListType tags;
+    check(xParameters::formatTagPrefix(tags).empty(),
+          "formatTagPrefix returns empty string for an empty tag list");
+}
+
+static void testHasTagGetTag() {
+    xParameters params;
+    xParameters::tagListType tags;
+    tags.push_back(MessageTag{"time", "2026-07-25T00:00:00.000Z"});
+    tags.push_back(MessageTag{"flag", ""});
+    params.setTags(tags);
+
+    check(params.hasTag("time"), "hasTag finds a present tag");
+    check(!params.hasTag("missing"), "hasTag returns false for an absent tag");
+
+    auto timeValue = params.getTag("time");
+    check(timeValue.has_value() && *timeValue == "2026-07-25T00:00:00.000Z",
+          "getTag returns the value of a present tag");
+
+    auto flagValue = params.getTag("flag");
+    check(flagValue.has_value() && flagValue->empty(),
+          "getTag returns an empty string_view for a value-less tag");
+
+    check(!params.getTag("missing").has_value(), "getTag returns nullopt for an absent tag");
+
+    params.Clear();
+    check(params.getTags().empty(), "Clear() also clears the tag list");
+}
+
+int main() {
+    testEscapeRoundTrip();
+    testUnescapeEdgeCases();
+    testParseTagsBasic();
+    testParseTagsMalformed();
+    testEscapedValueRoundTripThroughFormat();
+    testFormatTagPrefixEmpty();
+    testHasTagGetTag();
+
+    if (0 == failures) {
+        cout << "All xparameters tag tests passed." << endl;
+        return 0;
+    }
+
+    cout << failures << " test(s) failed." << endl;
+    return 1;
+}
```
